docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,554 @@
+            "$id": "#/properties/url",
+            "type": "string",
+            "format": "uri"
+          },
+          "additionalParametersSchema": { # JSON schema which describes additional properties passed in Capact workflow
+            "const": { # see http://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.6.1.3
+              "$id": "#/properties/additionalParameters",
+              "type": "object",
+              "required": [
+                "name",
+                "namespace"
+              ],
+              "properties": {
+                "name": {
+                  "$id": "#/properties/additionalParameters/properties/name",
+                  "type": "string"
+                },
+                "namespace": {
+                  "$id": "#/properties/additionalParameters/properties/namespace",
+                  "type": "string"
+                }
+              },
+              "additionalProperties": false
+            }
+          }
+        },
+        "additionalProperties": false
+      }
+    ```
+
+    **TODO:** How we can enforce such convention? Type composition?
+
+1. To install new Storage Backend, Cluster Admin has two options:
+
+    - use Capact Actions (e.g. `cap.interface.capactio.capact.hub.storage.helm-release.install`).
+    - Register a storage backend by creating such TypeInstance.
+
+    Regardless the option, at the end there is one TypeInstance produced:
+      
+      ```yaml
+      id: 3ef2e4ac-9070-4093-a3ce-142139fd4a16
+      metadata:
+        name: helm-storage
+      typeRef:
+        path: cap.type.helm.storage
+        revision: 0.1.0
+      latestResourceVersion:
+        metadata:
+          attributes:
+          - path: cap.core.attribute.hub.storage.backend # related to GrapHQL implementation
+            revision: 0.1.0
+        value:
+          url: "helm-release.default:50051"
+          additionalParametersSchema: {
+            "$id": "#/properties/additionalParametersSchema",
+            "type": "object",
+            "required": [
+              "name",
+              "namespace"
+            ],
+            "properties": {
+              "name": {
+                "$id": "#/properties/additionalParametersSchema/properties/name",
+                "type": "string"
+              },
+              "namespace": {
+                "$id": "#/properties/additionalParametersSchema/properties/namespace",
+                "type": "string"
+              }
+            },
+            "additionalProperties": false
+          }
+      backend: capact-redis # immutable - contains TypeInstance ID or unique alias
+        # if not provided, fallback to default one 
+      ```
+
+1. In fresh Capact installation, there is one TypeInstance already preregistered:
+
+    **TODO:** Would Redis be really the default storage? Probably PostgreSQL will be better, as we treat it as a "builtin" Backend as well (to store Attributes, relations, etc.)
```

Comment:
In past, I used Redis for such thing, but we really wanted sth fast. For relations, we created additional indexes. Those indexes were added and remove by Lua script running on Redis side.

IMO we don't need such optimization here, psql will be a good alternative 👍  Optionally we can start with a driver which will abstract used driver and if needed we can only replace it without changing our service logic.

Reply:
With many-to-many relations it would be inconvienent to have Redis. That's why I changed that to PostgreSQL today, during your review 👍